Do not generate default alt text for images

ckundo opened the pull request:
### Summary

- Deprecate `image_alt` helper.
- Remove the call to `image_alt` from `image_tag`.

[Fixes #30096]

### Impact

- After this change, screen readers will fallback to default, expected, and user configurable behavior for missing alt text.
- Also with this change, automated linting and testing tools will correctly generate warnings.

### Motivation

- Auto-generating content from the filename of an image is not suitable alternative text.
- Alt text that isn't fully considered can be distracting and fatiguing for screen readers users (blind, low vision, dyslexic people).
- Setting a filename fallback short circuits screen reader default behavior and configuration for blank descriptions, and may be frustrating to the assistive technology user.
- Setting poor defaults also creates false negatives for linting tools, that makes it harder to improve application accessibility.

ckundo commented:
I'm working on fixing up test failures now, but I gist of the PR is the same.

ckundo commented:
Something to reiterate from the PR comment: this change will create warnings in some accessibility testing tools. These would be _correct_ failures, but new failures nonetheless. In some cases these warnings could potentially be part of CI suites, and, in an even more narrow subset of applications, may create build failures.

Is it worth introducing a configuration option to disable the correct behavior for folks that may need to fix failures when upgrading Rails? The alternative for those that want to delay fixes would be to monkey-patch the `image_tag` helper in their apps until they remediate the issue. @mgifford how did y'all handle this in Drupal?

rafaelfranca commented on actionview/lib/action_view/helpers/asset_tag_helper.rb:
```diff
@@ -286,6 +282,8 @@ def image_tag(source, options = {})
       #   image_alt('underscored_file_name.png')
       #   # => Underscored file name
       def image_alt(src)
+        ::ActiveSupport::Deprecation.warn("ActionView::Helpers::AssesTagHelper.image_alt is deprecated and will be removed from Rails 5.2. You must exlicitly set alt text on images.")
```
I think you can remove the first `::` here.

rafaelfranca reviewed (changes requested):
It looks good to me. Nice work. Can you add a CHANGELOG entry?

georgeclaghorn commented on actionview/lib/action_view/helpers/asset_tag_helper.rb:
```diff
@@ -286,6 +282,8 @@ def image_tag(source, options = {})
       #   image_alt('underscored_file_name.png')
       #   # => Underscored file name
       def image_alt(src)
+        ActiveSupport::Deprecation.warn("ActionView::Helpers::AssesTagHelper.image_alt is deprecated and will be removed from Rails 5.2. You must exlicitly set alt text on images.")
```
`s/AssesTagHelper/AssetsTagHelper`, `s/exlicitly/explicitly`

rafaelfranca replied:
Don't need to talk about the module name too. Only `image_alt is deprecated` is good. And it is going to be removed from Rails 6.0.

georgeclaghorn commented on actionview/lib/action_view/helpers/asset_tag_helper.rb:
```diff
@@ -268,6 +262,8 @@ def image_tag(source, options = {})
         tag("img", options)
       end
 
+      # This method will be deprecated in future versions of Rails.
```
This method is deprecated now. It’ll be *removed* in Rails 5.2.

kaspth replied:
No, it's deprecated in 5.2 and will be removed in 6.0.

kaspth replied:
Though we let the deprecation warning itself suffice, so need for a comment here.

georgeclaghorn replied:
I meant that the comment says this method will eventually be deprecated, when it’s actually deprecated in this PR.

The deprecation warning says this method be removed in 5.2; it should say 6.0.

rafaelfranca reviewed (changes requested):
Looks good to me. Can you squash your commits?

ckundo committed:
Do not generate default alt text in image tags

- Auto-generating content from the filename of an image is not suitable
  alternative text; alt text that isn't fully considered can be
  distracting and fatiguing for screen readers users (blind, low vision,
  dyslexic people).
- Setting a filename fallback short circuits screen reader default
  behavior and configuration for blank descriptions.
- Setting poor defaults also creates false negatives for accessibility
  linting and testing software, that makes it harder to improve
  application accessibility.

***

- After this change, if authors leave images without alt text, screen
  readers will fallback to default behavior for missing alt text.
- Also with this change, Automated linting and testing tools will
  correctly generate warnings.

[Fixes #30096]

maclover7 commented on actionview/lib/action_view/helpers/asset_tag_helper.rb:
```diff
@@ -251,11 +249,7 @@ def image_tag(source, options = {})
-        unless src.start_with?("cid:") || src.start_with?("data:") || src.blank?
-          options[:alt] = options.fetch(:alt) { image_alt(src) }
-        end
+        options[:src] = path_to_image(source, skip_pipeline: skip_pipeline)
```
Why are we changing the method being used to generate `src`? Using `path_to_image` instead of `resolve_image_source` is causing CI to fail 😬

maclover7 commented on actionview/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+*   Remove default `alt` text generation
+
+    Fixes #30096
```
Can you add a period after the issue number

ckundo commented:
Thanks @maclover7, fixed. Good to squash?

rafaelfranca commented:
👍  squash again

ckundo commented:
Squashed and ready!

ckundo commented:
@rafaelfranca anything else I can add?
